### ISSUE TITLE
Does not work with lightgbm custom objective

### ISSUE THREAD

**Issue by elbaro** [dmlc/treelite#223]
Is Treelite compatible with lightgbm 3?
I am using a custom objective function and got `Ill-formed LightGBM model file: need objective`.

**Comment by hcho3** [comment]
@elbaro Yes, the latest development version supports the custom objective (#215)

**Comment by elbaro** [comment]
Can you reopen this?

The error comes from
```
 auto it = global_dict.find("objective");
    CHECK(it != global_dict.end())
      << "Ill-formed LightGBM model file: need objective";
```

I took a look at lightgbm 2 model file, and it has `objective=` field in the first few lines:
```
tree
version=..
num_class=..
objective=..
```

However in my lightgbm 3 model file, `objective=` is under `parameteres:`.
I guess this is why global_dict has no `objective` key.

**Comment by hcho3** [comment]
It appears that the `objective` field is left out when a custom objective is specified. Can you post a toy example with a custom objective in LightGBM? I'd like to add it to the test suite.

**Comment by elbaro** [comment]
```
def sigmoid(x):
    return 1 / (1+np.exp(-x))

# from https://github.com/Microsoft/LightGBM/blob/master/R-package/demo/early_stopping.R
def log_likelihood(preds, dataset):
    preds = sigmoid(preds)
    grad = preds - dataset.get_label()
    hess = preds * (1-preds)
    return grad, hess

def eval_metric(preds, dataset):
    preds = sigmoid(preds)
    err = ((preds > 0.5) != dataset.get_label()).mean()
    return err

model = lgb.train(
    config,
    train_dataset,
    valid_sets=[valid_dataset],
    valid_names=["validation"],
    fobj=log_likelihood,
    feval=eval_metric,
)
```

This is a python rewrite of the official lgb repo's R example.
I don't know why `evalerror` in the link assumed that the logistic transformation is already applied.

**Comment by elbaro** [comment]
This was not actually lightgbm 3 issue.

lightgbm sets `objective=None` when `fobj` is provided (3~4 years old code):
https://github.com/microsoft/LightGBM/blob/master/python-package/lightgbm/basic.py#L2437

And lightgbm's c++ part omits the objective field if it's none:
https://github.com/microsoft/LightGBM/blob/c07644d1d71540204a9b56f26667e8180bd009e2/src/boosting/gbdt_model_text.cpp#L320

By the way why does treelite need to know about objective function?

**Comment by hcho3** [comment]
This is so that the raw prediction is correctly transformed into the final prediction. For example, in binary classification setting, you'd want to transform raw prediction from the tree into a proper probability value by applying the softmax function.
